### PR TITLE
Feature/governance identity checks

### PR DESCRIPTION
**Pull request by JasonTulp**
Adding identity checks when adding a new council member
Adding Mock and test files

### CODE REVIEW

File: crml/governance/Cargo.toml [hunk 1]
```diff
@@ -11,9 +11,20 @@ description = "A runtime module for decentralized governance of the CENNZnet pro
...
 serde = { version = "1.0.102", optional = true }
 cennznet-primitives = { path = "../../primitives", default-features = false }
-frame-support = { git = "https://github.com/cennznet/substrate", rev = "d05b61e9b948621416a3c6eea2e10fd2f1b62aba", default-features = false }
-frame-system = { git = "https://github.com/cennznet/substrate", rev = "d05b61e9b948621416a3c6eea2e10fd2f1b62aba", default-features = false }
-sp-std = { git = "https://github.com/cennznet/substrate", rev = "d05b61e9b948621416a3c6eea2e10fd2f1b62aba", default-features = false }
+frame-support = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false }
+frame-system = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false }
+sp-std = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false }
+pallet-identity = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false }
```
Comment (jordy25519) [inline crml/governance/Cargo.toml]: should be in `dev-dependencies`

File: crml/governance/Cargo.toml [hunk 2]
```diff
@@ -11,9 +11,20 @@ description = "A runtime module for decentralized governance of the CENNZnet pro
...
+log = { version = "0.4.14", default-features = false }
+sp-runtime = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false  }
+sp-core = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false  }
+sp-staking = { git = "https://github.com/cennznet/substrate", default-features = false, rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b" }
```
Comment (jordy25519) [inline crml/governance/Cargo.toml]: unused?

File: crml/governance/Cargo.toml [hunk 3]
```diff
@@ -11,9 +11,20 @@ description = "A runtime module for decentralized governance of the CENNZnet pro
...
+sp-io = { git = "https://github.com/cennznet/substrate", rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b", default-features = false  }
+crml-support = { path = "../support", default-features = false }
+pallet-scheduler = { git = "https://github.com/cennznet/substrate", default-features = false, rev = "6ec19e279f8b47dabe39640ad93dfbf391d1e48b" }
```
Comment (jordy25519) [inline crml/governance/Cargo.toml]: another 'dev-dependencies' since its only required for tests

File: crml/governance/src/lib.rs [hunk 4]
```diff
@@ -53,6 +60,10 @@ pub trait Config: frame_system::Config {
 	type Event: From<Event> + Into<<Self as frame_system::Config>::Event>;
 	/// Weight information for extrinsics in this module.
 	type WeightInfo: WeightInfo;
+	/// Registrations for identities
+	type Registration: RegistrationInfo<AccountId = Self::AccountId>;
+
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: Add a doc comment

File: crml/governance/src/lib.rs [hunk 5]
```diff
@@ -103,6 +118,8 @@ decl_storage! {
 		NextProposalId get(fn next_proposal_id): ProposalId;
 		/// Proposal bond amount in 'wei'
 		ProposalBond get(fn proposal_bond): Balance;
+		/// Minimum stake required to create a new council member
+		MinimumCouncilStake get(fn minimum_council_stake): Balance = 20_000;
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: got to add the 0s to make it 4dp but lets leave this value as  `10_000_0000` to match the minimum stake default

File: crml/governance/src/lib.rs [hunk 6]
```diff
@@ -286,4 +322,23 @@ impl<T: Config> Module<T> {
 	pub fn get_proposal_votes() -> Vec<(ProposalId, ProposalVoteInfo)> {
 		ProposalVotes::iter().collect()
 	}
+
+	pub fn check_council_account_validity(account: &T::AccountId) -> DispatchResult {
+		// Check the amount they have staked
+		let staked_amount: Balance = T::StakingAmount::active_balance(account.clone());
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: nit: change this `active_balance` to take account by reference

File: crml/governance/src/lib.rs [hunk 7]
```diff
@@ -286,4 +322,23 @@ impl<T: Config> Module<T> {
...
+		frame_support::log::info!("The amount staked by this account is: {:?}", staked_amount);
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: don't need this log

File: crml/governance/src/lib.rs [hunk 8]
```diff
@@ -286,4 +322,23 @@ impl<T: Config> Module<T> {
 	pub fn get_proposal_votes() -> Vec<(ProposalId, ProposalVoteInfo)> {
 		ProposalVotes::iter().collect()
 	}
+
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: doc comment

File: crml/governance/src/lib.rs [hunk 9]
```diff
@@ -286,4 +322,23 @@ impl<T: Config> Module<T> {
...
+		let staked_amount: Balance = T::StakingAmount::active_balance(account.clone());
+		frame_support::log::info!("The amount staked by this account is: {:?}", staked_amount);
+
+		ensure!(
+			staked_amount >= Self::minimum_council_stake(),
+			Error::<T>::NotEnoughStaked
+		);
+
+		// Check their verified identities
+		let registration: u32 = T::Registration::registered_accounts(account.clone());
```
Comment (jordy25519) [inline crml/governance/src/lib.rs]: account by ref here too

Comment (jordy25519) [reply crml/governance/src/lib.rs]: nit: maybe rename `registered_accounts` to `registered_identity_count` so the term 'account' doesn't get confused here

File: crml/governance/src/tests.rs [hunk 10]
```diff
@@ -0,0 +1,91 @@
+use super::*;
+use crate::mock::{ExtBuilder, Governance, Test};
+use frame_support::{assert_noop, assert_ok};
+use sp_runtime::DispatchError;
+
+#[test]
```
Comment (jordy25519) [inline crml/governance/src/tests.rs:6]: nice tests :)

File: crml/governance/src/mock.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+pub struct MockStakingAmount;
+impl StakingAmount for MockStakingAmount {
+	type AccountId = AccountId;
+	type Balance = Balance;
+
+	fn active_balance(controller: Self::AccountId) -> Self::Balance {
+		match controller {
+			1 => 10000,
+			_ => 20000,
```
Comment (jordy25519) [inline crml/governance/src/mock.rs]: would make missing accounts 0 so you don't get confusing test results
also needs the extra 4 `0`s
```suggestion
			1 => 100000000,
			2 => 200000000,
			_ => 0,
```

File: crml/governance/src/mock.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+pub struct MockRegistrationImplementation;
+impl RegistrationInfo for MockRegistrationImplementation {
+	type AccountId = AccountId;
+	/// Registration information for an identity
+	fn registered_accounts(who: Self::AccountId) -> u32 {
+		match who {
+			1 => 2,
+			2 => 1,
+			_ => 2,
```
Comment (jordy25519) [inline crml/governance/src/mock.rs]: same here catchall case `_` should be 0
```suggestion
			1 => 2,
			2 => 1,
			_ => 0,
```

File: crml/governance/Cargo.toml (hunk 1 above)
Comment (JasonTulp) [reply crml/governance/Cargo.toml]: Actually unused

File: crml/governance/Cargo.toml (hunk 3 above)
Comment (JasonTulp) [reply crml/governance/Cargo.toml]: This is required by runtime so I think needs to stay where it is

**Review by jordy25519 — APPROVED** [review]

**Comment by jordy25519** [comment]
just needs merge conflicts resolved

**Pull request merged**